### ISSUE TITLE
amplify add auth breaks entire project

### ISSUE THREAD

**Issue by ernestostifano** [aws-amplify/amplify-cli#9363]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

17.2.0

### Amplify CLI Version

7.6.4

### What operating system are you using?

macOS 10.15.7

### Amplify Categories

auth

### Amplify Commands

add

### Describe the bug

After running `amplify add auth` and following all steps of manual configuration (see complete log below) now all the project is broken. Every time I run an Amplify command I get the following error: `Name: TypeError : Message: Cannot read properties of undefined (reading 'serviceType')`

### Expected behavior

Auth category to be correctly added without breaking anything.

### Reproduction steps

See logs below.

### GraphQL schema(s)

N/A

### Log output

<details>

```
# Put your logs below this line
esmbp1015:codistica-website ernestostifano$ amplify add auth
Using service: Cognito, provided by: awscloudformation
 
 The current configured provider is Amazon Cognito. 
 
 Do you want to use the default authentication and security configuration? Manual configuration
 Select the authentication/authorization services that you want to use: User Sign-Up, Sign-In, connected with AWS IAM controls (Enables per-user Storage features for images or
 other content, Analytics, and more)
 Provide a friendly name for your resource that will be used to label this category in the project: CodisticaWebsiteAuth
 Enter a name for your identity pool. CodisticaWebsiteIdentityPool
 Allow unauthenticated logins? (Provides scoped down permissions that you can control via AWS IAM) Yes
 Do you want to enable 3rd party authentication providers in your identity pool? No
 Provide a name for your user pool: CodisticaWebsiteUserPool
 Warning: you will not be able to edit these selections. 
 How do you want users to be able to sign in? Email
 Do you want to add User Pool Groups? Yes
? Provide a name for your user pool group: admins
? Do you want to add another User Pool Group Yes
? Provide a name for your user pool group: customers
? Do you want to add another User Pool Group No
✔ Sort the user pool groups in order of preference · admins, customers
 Do you want to add an admin queries API? Yes
? Do you want to restrict access to the admin queries API to a specific Group Yes
? Select the group to restrict access with: admins
 Multifactor authentication (MFA) user login options: OPTIONAL (Individual users can use MFA)
 For user login, select the MFA types: Time-Based One-Time Password (TOTP)
 Specify an SMS authentication message: Your authentication code is {####}
 Email based user registration/forgot password: Enabled (Requires per-user email entry at registration)
 Specify an email verification subject: Your verification code
 Specify an email verification message: Your verification code is {####}
 Do you want to override the default password policy for this User Pool? Yes
 Enter the minimum password length for this User Pool: 8
 Select the password character requirements for your userpool: Requires Lowercase, Requires Uppercase, Requires Numbers, Requires Symbols
 Warning: you will not be able to edit these selections. 
 What attributes are required for signing up? Email
 Specify the app's refresh token expiration period (in days): 30
 Do you want to specify the user attributes this app can read and write? No
 Do you want to enable any of the following capabilities? Email Verification Link with Redirect, Add User to Group
 Do you want to use an OAuth flow? No
? Do you want to configure Lambda Triggers for Cognito? No
2021-12-21T12:42:16.049Z - error: uncaughtException: Cannot read properties of undefined (reading 'push') date=Tue Dec 21 2021 13:42:16 GMT+0100 (Central European Standard Time), pid=9538, uid=501, gid=20, cwd=/Users/ernestostifano/repositories/codistica-website, execPath=/usr/local/Cellar/node/17.2.0/bin/node, version=v17.2.0, argv=[/usr/local/Cellar/node/17.2.0/bin/node, /usr/local/bin/amplify, add, auth], rss=597618688, heapTotal=533700608, heapUsed=494870648, external=3239868, arrayBuffers=1300582, loadavg=[1.3837890625, 1.638671875, 1.560546875], uptime=14176, trace=[column=38, file=/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts, function=AmplifyAuthCognitoStack.generateCognitoStackResources, line=382, method=generateCognitoStackResources, native=false, column=27, file=/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-stack-transform.ts, function=AmplifyAuthTransform.generateStackResources, line=95, method=generateStackResources, native=false, column=16, file=/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-stack-transform.ts, function=AmplifyAuthTransform.transform, line=69, method=transform, native=false, column=10, file=/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-stack-template.ts, function=generateAuthStackTemplate, line=6, method=null, native=false, column=7, file=/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/provider-utils/awscloudformation/handlers/resource-handlers.ts, function=null, line=68, method=null, native=false, column=3, file=/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/index.js, function=Object.executeAmplifyCommand, line=421, method=executeAmplifyCommand, native=false, column=3, file=/usr/local/lib/node_modules/@aws-amplify/cli/src/execution-manager.ts, function=executePluginModuleCommand, line=204, method=null, native=false, column=5, file=/usr/local/lib/node_modules/@aws-amplify/cli/src/execution-manager.ts, function=executeCommand, line=30, method=null, native=false, column=5, file=/usr/local/lib/node_modules/@aws-amplify/cli/src/index.ts, function=Object.run, line=205, method=run, native=false], stack=[TypeError: Cannot read properties of undefined (reading 'push'),     at AmplifyAuthCognitoStack.generateCognitoStackResources (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:382:38),     at AmplifyAuthTransform.generateStackResources (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-stack-transform.ts:95:27),     at AmplifyAuthTransform.transform (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-stack-transform.ts:69:16),     at generateAuthStackTemplate (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-stack-template.ts:6:10),     at /usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/provider-utils/awscloudformation/handlers/resource-handlers.ts:68:7,     at Object.executeAmplifyCommand (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/index.js:421:3),     at executePluginModuleCommand (/usr/local/lib/node_modules/@aws-amplify/cli/src/execution-manager.ts:204:3),     at executeCommand (/usr/local/lib/node_modules/@aws-amplify/cli/src/execution-manager.ts:30:5),     at Object.run (/usr/local/lib/node_modules/@aws-amplify/cli/src/index.ts:205:5)]
esmbp1015:codistica-website ernestostifano$ amplify add auth
Auth has already been added to this project. To update run amplify update auth.
esmbp1015:codistica-website ernestostifano$ amplify update auth
Please note that certain attributes may not be overwritten if you choose to use defaults settings.
Cannot read properties of undefined (reading 'serviceType')
TypeError: Cannot read properties of undefined (reading 'serviceType')
    at getResourcesToBeCreated (/usr/local/lib/node_modules/@aws-amplify/cli/src/extensions/amplify-helpers/resource-status-data.ts:220:65)
    at AmplifyToolkit.getResourceStatus [as _getResourceStatus] (/usr/local/lib/node_modules/@aws-amplify/cli/src/extensions/amplify-helpers/resource-status-data.ts:117:35)
    at getAuthResourceName (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/utils/getAuthResourceName.ts:4:50)
    at Object.run (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/commands/auth/update.ts:58:49)
    at Object.executeAmplifyCommand (/usr/local/lib/node_modules/@aws-amplify/cli/node_modules/@aws-amplify/amplify-category-auth/src/index.js:421:23)
    at /usr/local/lib/node_modules/@aws-amplify/cli/src/execution-manager.ts:222:45
    at executePluginModuleCommand (/usr/local/lib/node_modules/@aws-amplify/cli/src/execution-manager.ts:204:9)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at executeCommand (/usr/local/lib/node_modules/@aws-amplify/cli/src/execution-manager.ts:30:5)
    at Object.run (/usr/local/lib/node_modules/@aws-amplify/cli/src/index.ts:205:5)
esmbp1015:codistica-website ernestostifano$ amplify status
Name: TypeError : Message: Cannot read properties of undefined (reading 'serviceType')
esmbp1015:codistica-website ernestostifano$ 


```

</details>

### Additional information

This was the first category to be added, so project was "fresh", only CI/CD hosting was present and correctly working.

**Comment by queryluke** [comment]
I've been having a nearly similar issue. Fresh project, auth is the first thing I've added, same CLI version.

Except my error is `uncaughtException: Cannot read properties of undefined (reading 'push')` (but I've been trying to push the newly created auth)

What I've found so far is that running `amplify auth add` and choosing default configuration (first choice) removes the error. But running manual configuration with my preferred options causes the error.

For me, I isolated the issue to be the `Multifactor authentication (MFA) user login options:`, specifically the `For user login, select the MFA types:` . In short, it appears selecting `Time-Based One-Time Password (TOTP)` causes the error.

If you were selecting that, it could be the cause. Otherwise another option (which I haven't selected) might be the issue.

Hope that helps a little.

**Comment by ernestostifano** [comment]
@queryluke thank you very much for taking the time to write your response. I'm sure it will help a lot. I was already thinking of downgrading Amplify CLI and/or Node.

Yes, I selected `Time-Based One-Time Password (TOTP)`. Later today I will try to downgrade Amplify CLI, setup Auth and then upgrade again if needed. I'll let you know how it goes. Worst case, I'll choose SMS as MFA method until this issue is solved. Hope there is not another option in my manual config that will make process crash again.

Thanks again.

**Comment by ernestostifano** [comment]
### Update
- `Node@17.2.0` + `@aws-amplify/cli@7.6.5` -> ERROR
- `Node@17.2.0` + `@aws-amplify/cli@7.6.4` -> ERROR
- `Node@17.2.0` + `@aws-amplify/cli@7.5.6` -> ERROR
- `Node@16.13.1` + `@aws-amplify/cli@7.6.5` -> ERROR

**Comment by ernestostifano** [comment]
## Update

`node@17.2.0` + `@aws-amplify/cli7.6.5`

If I select `SMS Text Message` AND `Time-Based One-Time Password (TOTP)` there is no error!

**Comment by danielleadams** [comment]
Does this workflow work with Node 14? While I understand Node 17 is an actively worked on and supported version of Node, I would recommend using @aws-amplify/cli with Node 14 or 16.

**Comment by josefaidt** [comment]
Hey @ernestostifano and @queryluke thanks for providing those details! Using Node 16 and the latest version of the Amplify CLI I was able to successfully reproduce by selecting either TOTP or SMS when configuring MFA for an auth resource, but as you have also noticed if we select both options we do not experience the issue. The issue also occurs when updating an existing auth resource

```
> amplify update auth
Please note that certain attributes may not be overwritten if you choose to use defaults settings.
Using service: Cognito, provided by: awscloudformation
 What do you want to do? Walkthrough all the auth configurations
 Select the authentication/authorization services that you want to use: User Sign-Up, Sign-In, connected with AWS IA
M controls (Enables per-user Storage features for images or other content, Analytics, and more)
 Allow unauthenticated logins? (Provides scoped down permissions that you can control via AWS IAM) No
 Do you want to enable 3rd party authentication providers in your identity pool? No
 Do you want to add User Pool Groups? No
 Do you want to add an admin queries API? No
 Multifactor authentication (MFA) user login options: OPTIONAL (Individual users can use MFA)
 For user login, select the MFA types: Time-Based One-Time Password (TOTP)
 Specify an SMS authentication message: Your authentication code is {####}
 Email based user registration/forgot password: Enabled (Requires per-user email entry at registration)
 Specify an email verification subject: Your verification code
 Specify an email verification message: Your verification code is {####}
 Do you want to override the default password policy for this User Pool? No
 Specify the app's refresh token expiration period (in days): 30
 Do you want to specify the user attributes this app can read and write? No
 Do you want to enable any of the following capabilities? 
 Do you want to use an OAuth flow? No
? Do you want to configure Lambda Triggers for Cognito? No
```

One difference I noticed during my reproduction was the stack trace is pointing to a different location, https://github.com/aws-amplify/amplify-cli/blob/master/packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts#L395

Finally, after successfully updating the auth resource by choosing both MFA options, we are no longer able to update MFA by following the `amplify update auth` manual walkthrough again although we did not yet push the update.

Marking as a bug 🙂

**Comment by attilah** [comment]
The side effect you are seeing is fixed in #9214 that will be in the next release. The root case for the `serviceType` related failure is that UserPool Groups resource is already added to the `backend-config.json` and `amplify-meta.json` and it references the non-existent `auth` Cognito resource.

I'm creating a fix to not to fail in this specific scenario, but on the meantime you can unblock yourself by removing the `userPoolGroups` resource from the above mentioned files within `amplify/backend` folder and also remove the empty object under `auth` in `team-provider-info.json`.

Let me know if that helped to unblock you to not have a "corrupt" project.
